include/psa/crypto_extra.h

```diff
@@ -813,6 +813,987 @@ psa_status_t mbedtls_psa_platform_get_builtin_key(
 
 /** @} */
 
+/** \addtogroup crypto_types
+ * @{
+ */
+
+#define PSA_ALG_CATEGORY_PAKE                   ((psa_algorithm_t)0x0a000000)
+
+/** Whether the specified algorithm is a password-authenticated key exchange.
+ *
+ * \param alg An algorithm identifier (value of type #psa_algorithm_t).
+ *
+ * \return 1 if \p alg is a password-authenticated key exchange (PAKE)
+ *         algorithm, 0 otherwise.
+ *         This macro may return either 0 or 1 if \p alg is not a supported
+ *         algorithm identifier.
+ */
+#define PSA_ALG_IS_PAKE(alg)                                        \
+    (((alg) & PSA_ALG_CATEGORY_MASK) == PSA_ALG_CATEGORY_PAKE)
+
+/** The Password-authenticated key exchange by juggling (J-PAKE) algorithm.
+ *
+ * This is J-PAKE as defined by RFC 8236, instantiated with the following
+ * parameters:
+ *
+ * - The group can be either an elliptic curve or defined over a finite field.
+ * - Schnorr NIZK proof as defined by RFC 8235 and using the same group as the
+ *   J-PAKE algorithm.
+ * - A secure cryptographic hash function.
+ *
+ * To select these parameters and set up the cipher suite, call
+ *      psa_pake_cs_set_algorithm(cipher_suite, PSA_ALG_JPAKE);
+ *      psa_pake_cs_set_primitive(cipher_suite,
+ *                                PSA_PAKE_PRIMITIVE(type, family, bits));
+ *      psa_pake_cs_set_hash(cipher_suite, hash);
+ *
+ * For more information on how to set a specific curve or field, refer to the
+ * documentation of the individual \c PSA_PAKE_PRIMITIVE_TYPE_XXX constants.
+ *
+ * After initializing a J-PAKE operation, call
+ *      psa_pake_setup(operation, cipher_suite);
+ *      psa_pake_set_user(operation, ...);
+ *      psa_pake_set_peer(operation, ...);
+ *      psa_pake_set_password_key(operation, ...);
+ *
+ * The password is read as a byte array and must be non-empty. This can be the
```

Comment:
Why would an empty password not be supported? I don't have a concrete example, but it seems plausible that a protocol with an optional password would have a mandatory PAKE to keep the software design simple, and would use an empty password if the security policy said that a password is not used.

A precedent for allowing an empty secret input is in key derivation, with a concrete use case of using the KDF algorithm to derive both a key and an IV, with an empty “secret” input to the KDF when deriving the IV. For this case, `psa_key_derivation_input_key` can't be used because keys never have zero length, but `psa_key_derivation_input_bytes` does accept an empty input.

Reply:
This might be because J-PAKE specifically requires the secret  _s_ &ne; 0? For J-PAKE, the secret is an integer - and a common conversion from a byte-string to an integer results in the value zero for the empty string.

Reply:
Good point, makes sense, so you can't to this with J-PAKE. But I think this should be a constraint for `PSA_ALG_JPAKE` rather than a mandatory constraint for all algorithms.

Reply:
J-PAKE does not prescribe how we make an integer of the password though. If the application or higher level protocol chooses to hash the password and convert it to an integer, then empty passwords become feasible.

(If the secret integer `s` is 0 mod the group order, then the key material will always be the generator element.)

Reply:
> But I think this should be a constraint for `PSA_ALG_JPAKE` rather than a mandatory constraint for all algorithms.

I agree. As far as I can tell, this part is specific to `PSA_ALG_JPAKE` and the description of `psa_pake_set_password_key()` doesn't mention such restrictions.

Reply:
Given that J-PAKE expects the secret to be a big integer, and the PAKE API only accepts a key (which is typically considered to be an octet string) as input - this specification needs to document how the key material is converted to the integer _s_.

That conversion is not specified by J-PAKE, it only hints at some application options. As this is a two party protocol, and we shouldn't assume PSA Crypto at the other end, we need to enable any reasonable approach. That suggests that we should use the most obvious "octet-string is the big*-endian representation of an integer" (* or little?), which allows application flexibility (to hash or not to hash).

Reply:
It is documented in the next paragraph:
https://github.com/ARMmbed/mbedtls/pull/4317/files#diff-8fe19c51d861f48010b3609673ee1dd94bb3eef20af920af31b36fc0dcb124c8R863-R868

Reply:
(This allows for an application to do hashing beforehand if they wish or any other transformation that is required.)